### ISSUE TITLE
doing more with less

### ISSUE THREAD

**Issue by RobinTournemenne** [anthonyalfimov/SWONIC-Dynamic-Controls#63]
Very inspiring work!

I want to push the boundaries even further having simply 25 buttons whose aftertouch in the 4 cartesian direction would trigger different actions (could be XY control, fadder or simply demultiply each button in 4 buttons). It seems pretty straightforward thanks to the touchMove built-in function.

In the end, it will be only be a "redesign" of your neat application, but will let the user have even more control on this tiny control pad.

**Comment by RobinTournemenne** [comment]
I made it! It was pretty tedious. Littlefoot is such an unfinished scripting language... between the buggy IDE, the memory hassle and the lack of documenation... Your silent help through your scripts and your piece of advise on the JUCE forum was more than welcome!

Here are the files:

[littlefootandmode.zip](https://github.com/anthonyalfimov/SWONIC-Dynamic-Controls/files/4694911/littlefootandmode.zip)

This is based on the control grid factory program of ROLI. It is a matrix of 25 buttons which can either be a note, a simple CC, a 3-ways CC or a fadder.

If the button is a 3-way CC, once you've pressed it if you go down or up it will trigger after some distance a CC (I only really thought of the gate type of control since I don't really use the others, but I left the part of the initial program dealing with togle and trigger).

If it is a fadder, after pressing the button, the up and down aftertouch movement will set the fadder value.

There are also some magical CC number:
 CC == 128 // previous octave
 CC == 129 // next octave
 CC == 130 // previous minor third
 CC == 131 // next minor third
 CC == 132 // toggle fixed velocity
 CC == 133 // nothing happens
 CC == 134 // fadder for velocity choice

I also attached the mode I am currently using for work to get an idea of how I organised the pad. After playing with it for some time I can definitely tell I will use it very often instead of my fully fledged keyboard!

**Comment by anthonyalfimov** [comment]
This is a very interesting idea, thanks for sharing it! 

As I understand, one of the core ideas here is to get as many control inputs as possible from a single Lightpad. But I think this concept could also work really well with a 3 x 3 grid of 5 x 5 px controls (at least as a pure CC controller). With the current setup, you can't really do a "down" motion on the bottom row of the pads. But with bigger pads, you can perform (and visualise!) all the motions within the constraints of the pad itself. You could even add different pressure zones on a single pad.

**Comment by SWONIC** [comment]
@RobinTournemenne yes, thank you for sharing this! :-)

Actually, that was also my experience when testing it. That not all motions make sense on all positions of the Lightpad. It's actually also on of the biggest weaknesses of the 5x5 Note Grid app by ROLI - sliding and gliding actually really works for the centered controls... I like @anthonyalfimov idea with the 3x3 grid. But if you want to maximise the possible controller inputs the 5x5 grid might be a better option.

Another thing you might want to add is manual colouring of the controls? Having them in random colour might be a bit confusing if someone would try to create a template for their setup...

**Comment by RobinTournemenne** [comment]
I think too that cramping buttons on 1 lightpad has nothing else than an interest for controls only. It may be possible to get back some expresiveness introducing a small delay for the artist to reach to the center of the pad before starting some sound modification using aftertouch. Your strategy is also a good idea, while going into the category "doing a lot more with more" (the context for me is working on the go, so one lightpad in bluetooth only).

Since yesterday, I improved the fadder controls using coarse positionning with the Y axis and fine tuning with the X axis. It improved a lot the fadder stability issues I got (in my workflow, it is important to reach very specific midi CC values like 49, 74 or 99). I realize now that this two axes strategies can be used to get other types of value selections.

Regarding colors... I removed the color choice in order to get back some memory! I also wanted to create 5 ways buttons (x-, x+, y-, y+, press) but couldn't because of memory struggle................. If you have any idea to improve those issues, it is more than welcome. My idea right now is to tailor the available possibilities for each button (e.g. the button at the bottom left can only do x+, y+, press so I don't define x- and y- useless variables), and define afterwards getters and setters (which will eat memory once again........)

Here is the new version of my code:

[littlefootandmode_2.zip](https://github.com/anthonyalfimov/SWONIC-Dynamic-Controls/files/4705297/littlefootandmode_2.zip)

Thanks a lot for your feedback!
